### ISSUE TITLE
[interface-name-prefix] <rule doesn't behave as expected>

### ISSUE THREAD

**Issue by desirelabs** [typescript-eslint/typescript-eslint#745]
Hello,

[interface-name-prefix] plugin doesn't return any error when set to `2` (error) when en interface is not prefixed.

```JSON
{
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "extends": ["plugin:@typescript-eslint/recommended"],
  "rules": {
    "indent": "off",
    "@typescript-eslint/indent": ["error", 2],
    "@typescript-eslint/no-explicit-any": 2,
    "@typescript-eslint/no-unused-vars": 2,
    "@typescript-eslint/interface-name-prefix": 2
  }
}
```
Code example :
```TS
interface Props {
  name: string;
  age: number;
}

function test(props: Props): Props {
  return props
};

test({name: "foo", age: 42})
```

**Expected Result**
Eslint should return an error on `Props` interface not being prefixed with an `I`.

**Actual Result**
Eslint doesn't return any error.

**debug dump**
```shell
  eslint:cli CLI args: [ './index.ts', '--debug' ] +0ms
  eslint:cli Running on files +6ms
  eslint:ignored-paths baseDir = "/Users/franck/workspace/lab/typescript-eslint" +0ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "/node_modules/*"
  eslint:ignored-paths   cooked   = "/node_modules/*" +1ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "/bower_components/*"
  eslint:ignored-paths   cooked   = "/bower_components/*" +1ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = ".*"
  eslint:ignored-paths   cooked   = ".*" +0ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "!../"
  eslint:ignored-paths   cooked   = "!../" +0ms
  eslint:ignored-paths Looking for ignore file in /Users/franck/workspace/lab/typescript-eslint +0ms
  eslint:ignored-paths Could not find ignore file in cwd +0ms
  eslint:ignored-paths baseDir = "/Users/franck/workspace/lab/typescript-eslint" +1ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "/node_modules/*"
  eslint:ignored-paths   cooked   = "/node_modules/*" +1ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "/bower_components/*"
  eslint:ignored-paths   cooked   = "/bower_components/*" +0ms
  eslint:ignored-paths Looking for ignore file in /Users/franck/workspace/lab/typescript-eslint +0ms
  eslint:ignored-paths Could not find ignore file in cwd +0ms
  eslint:file-enumerator Start to iterate files: [ './index.ts' ] +0ms
  eslint:file-enumerator File: /Users/franck/workspace/lab/typescript-eslint/index.ts +1ms
  eslint:cascading-config-array-factory Load config files for /Users/franck/workspace/lab/typescript-eslint. +0ms
  eslint:cascading-config-array-factory No cache found: /Users/franck/workspace/lab/typescript-eslint. +0ms
  eslint:config-array-factory Loading legacy config file: /Users/franck/workspace/lab/typescript-eslint/.eslintrc +0ms
  eslint:config-array-factory Config file found: /Users/franck/workspace/lab/typescript-eslint/.eslintrc +40ms
  eslint:config-array-factory Loading {extends:"plugin:@typescript-eslint/recommended"} relative to /Users/franck/workspace/lab/typescript-eslint/.eslintrc +21ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from /Users/franck/workspace/lab/typescript-eslint/.eslintrc +0ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@1.13.0 (/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/index.js) +4ms
  eslint:config-array-factory Loading {extends:"./configs/base.json"} relative to /Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/index.js +282ms
  eslint:config-array-factory package.json was not found: Cannot find module './configs/base.json/package.json'
Require stack:
- /Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/index.js +1ms
  eslint:config-array-factory Loaded: ./configs/base.json (/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.json) +0ms
  eslint:config-array-factory Loading JSON config file: /Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.json +0ms
  eslint:config-array-factory Loading parser "@typescript-eslint/parser" from /Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.json +0ms
  eslint:config-array-factory Loaded: @typescript-eslint/parser@1.13.0 (/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/parser/dist/parser.js) +1ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from /Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.json +25ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@1.13.0 (/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/index.js) +0ms
  eslint:config-array-factory Loading parser "@typescript-eslint/parser" from /Users/franck/workspace/lab/typescript-eslint/.eslintrc +1ms
  eslint:config-array-factory Loaded: @typescript-eslint/parser@1.13.0 (/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/parser/dist/parser.js) +0ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from /Users/franck/workspace/lab/typescript-eslint/.eslintrc +0ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@1.13.0 (/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/index.js) +1ms
  eslint:cascading-config-array-factory No cache found: /Users/franck/workspace/lab. +376ms
  eslint:config-array-factory Config file not found on /Users/franck/workspace/lab +0ms
  eslint:cascading-config-array-factory No cache found: /Users/franck/workspace. +0ms
  eslint:config-array-factory Config file not found on /Users/franck/workspace +0ms
  eslint:cascading-config-array-factory No cache found: /Users/franck. +0ms
  eslint:cascading-config-array-factory Stop traversing because of considered root. +0ms
  eslint:rules Loading rule 'camelcase' (remaining=265) +0ms
  eslint:rules Loading rule 'indent' (remaining=264) +7ms
  eslint:rules Loading rule 'no-array-constructor' (remaining=263) +17ms
  eslint:rules Loading rule 'no-unused-vars' (remaining=262) +9ms
  eslint:rules Loading rule 'no-use-before-define' (remaining=261) +2ms
  eslint:cascading-config-array-factory Configuration was determined: ConfigArray [ { name: '.eslintrc » plugin:@typescript-eslint/recommended » ./configs/base.json', filePath: '/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.json', criteria: null, env: undefined, globals: undefined, parser: { error: null, filePath: '/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/parser/dist/parser.js', id: '@typescript-eslint/parser', importerName: '.eslintrc » plugin:@typescript-eslint/recommended » ./configs/base.json', importerPath: '/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.json' }, parserOptions: { sourceType: 'module' }, plugins: { '@typescript-eslint': [Object] }, processor: undefined, root: undefined, rules: undefined, settings: undefined }, { name: '.eslintrc » plugin:@typescript-eslint/recommended', filePath: '/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/eslint-plugin/dist/index.js', criteria: null, env: undefined, globals: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, root: undefined, rules: { '@typescript-eslint/adjacent-overload-signatures': 'error', '@typescript-eslint/array-type': 'error', '@typescript-eslint/ban-types': 'error', camelcase: 'off', '@typescript-eslint/camelcase': 'error', '@typescript-eslint/class-name-casing': 'error', '@typescript-eslint/explicit-function-return-type': 'warn', '@typescript-eslint/explicit-member-accessibility': 'error', indent: 'off', '@typescript-eslint/indent': 'error', '@typescript-eslint/interface-name-prefix': 'error', '@typescript-eslint/member-delimiter-style': 'error', '@typescript-eslint/no-angle-bracket-type-assertion': 'error', 'no-array-constructor': 'off', '@typescript-eslint/no-array-constructor': 'error', '@typescript-eslint/no-empty-interface': 'error', '@typescript-eslint/no-explicit-any': 'warn', '@typescript-eslint/no-inferrable-types': 'error', '@typescript-eslint/no-misused-new': 'error', '@typescript-eslint/no-namespace': 'error', '@typescript-eslint/no-non-null-assertion': 'error', '@typescript-eslint/no-object-literal-type-assertion': 'error', '@typescript-eslint/no-parameter-properties': 'error', '@typescript-eslint/no-triple-slash-reference': 'error', 'no-unused-vars': 'off', '@typescript-eslint/no-unused-vars': 'warn', 'no-use-before-define': 'off', '@typescript-eslint/no-use-before-define': 'error', '@typescript-eslint/no-var-requires': 'error', '@typescript-eslint/prefer-interface': 'error', '@typescript-eslint/prefer-namespace-keyword': 'error', '@typescript-eslint/type-annotation-spacing': 'error' }, settings: undefined }, { name: '.eslintrc', filePath: '/Users/franck/workspace/lab/typescript-eslint/.eslintrc', criteria: null, env: undefined, globals: undefined, parser: { error: null, filePath: '/Users/franck/workspace/lab/typescript-eslint/node_modules/@typescript-eslint/parser/dist/parser.js', id: '@typescript-eslint/parser', importerName: '.eslintrc', importerPath: '/Users/franck/workspace/lab/typescript-eslint/.eslintrc' }, parserOptions: undefined, plugins: { '@typescript-eslint': [Object] }, processor: undefined, root: undefined, rules: { indent: 'off', '@typescript-eslint/indent': [Array], '@typescript-eslint/no-explicit-any': 2, '@typescript-eslint/no-unused-vars': 2, '@typescript-eslint/interface-name-prefix': 2 }, settings: undefined } ] on /Users/franck/workspace/lab/typescript-eslint +48ms
  eslint:ignored-paths contains: +429ms
  eslint:ignored-paths   target   = "/Users/franck/workspace/lab/typescript-eslint/index.ts" +0ms
  eslint:ignored-paths   base     = "/Users/franck/workspace/lab/typescript-eslint" +0ms
  eslint:ignored-paths   relative = "index.ts" +0ms
  eslint:ignored-paths   result   = false +0ms
  eslint:cli-engine Lint /Users/franck/workspace/lab/typescript-eslint/index.ts +0ms
  eslint:linter Linting code for /Users/franck/workspace/lab/typescript-eslint/index.ts (pass 1) +0ms
  eslint:linter Verify +1ms
  eslint:linter With ConfigArray: /Users/franck/workspace/lab/typescript-eslint/index.ts +0ms
  eslint:linter Generating fixed text for /Users/franck/workspace/lab/typescript-eslint/index.ts (pass 1) +76ms
  eslint:source-code-fixer Applying fixes +0ms
  eslint:source-code-fixer shouldFix parameter was false, not attempting fixes +0ms
  eslint:file-enumerator Complete iterating files: ["./index.ts"] +505ms
  eslint:cli-engine Linting complete in: 507ms +78ms
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^1.13.0` |
| `@typescript-eslint/parser`        | `^1.13.0` |
| `TypeScript`                       | `^3.5.3` |
| `ESLint`                           | `^6.1.0` |
| `node`                             | `10.16.0` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
you've got the rule backwards. The default config is never prefix a rule with an I, as it is an antipattern that shouldn't be used.

If you want to use the antipattern, you must configure the rule. See the rule doc for more info.

**Comment by desirelabs** [comment]
I'm sorry, yes I want to use the antipattern. And the rule doc tells this rule has a string option, in our case, "always". The typescript rule tells to use something like `"interface-name": [true, "always-prefix"]`.

It appears that the option "always" for the rule `@typescript-eslint/interface-name-prefix` is not working as expected, and the linter yiels that it should be a number.

This is a bit confusing and there is no proper example about how to implement it the right way. So IMHO, it's clear that I have a lack of knowledge, and apart than the fact that many people says it's and antipattern, I think it would be constructive to give more assets to understand and implement this rule.

I'm very sorry, but this doesn't help me much at the moment. Please, consider giving some more tips.

Cheers

**Comment by bradzacher** [comment]
https://eslint.org/docs/user-guide/configuring#configuring-rules

```
"@typescript-eslint/interface-name-prefix": ["error", "always-prefix"]
```

**Comment by IslamWahid** [comment]
@bradzacher this is not working also

**Comment by IslamWahid** [comment]
just found that it should be like this 
```
"@typescript-eslint/interface-name-prefix": [
      "error",
      "always"
    ]
```

**Comment by bradzacher** [comment]
that would be because the rule has been updated since I wrote that comment...

The best place to learn about how to configure rules is via the rule's documentation.

**Comment by IslamWahid** [comment]
@bradzacher the problem is that rule's documentation isn't updated also

**Comment by bradzacher** [comment]
Are you sure? It looks like it's documented:

https://github.com/typescript-eslint/typescript-eslint/blob/0c4f474/packages/eslint-plugin/docs/rules/interface-name-prefix.md#L14-L26

![image](https://user-images.githubusercontent.com/7462525/64925505-51e4c180-d7a6-11e9-8475-4928986257a0.png)

**Comment by IslamWahid** [comment]
@bradzacher well thank you

**Comment by iamdevlinph** [comment]
The rule does not work with the [documented options](https://github.com/typescript-eslint/typescript-eslint/blob/v2.12.0/packages/eslint-plugin/docs/rules/interface-name-prefix.md)

![image](https://user-images.githubusercontent.com/19337229/71060963-eb56b180-21a1-11ea-9244-52dfaa53027f.png)

![image](https://user-images.githubusercontent.com/19337229/71060970-edb90b80-21a1-11ea-8eba-3dfced945d5c.png)

But it works when following the eslint way.

![image](https://user-images.githubusercontent.com/19337229/71061091-3c66a580-21a2-11ea-8cca-970e276a37d1.png)

**Comment by bradzacher** [comment]
> am I missing something

You are.
Eslint rules are configured the same way.

```
"ruleName": ["error", <ruleConfig>]
```

In your first example, you're not configuring it correctly at all. If you ran a cli lint run, you'd probably find eslint would barf at you.

Try putting the object as a second item to the array, instead of the string...
